feat: run instance creation in parallel

4censord opened the pull request:
Basically just changes the loops in Increase and Decrease to run in parallel using go routines.

So now instead of waiting until one instance is completely created on the cloudscale side (`Servers.WaitFor`) and only then continuing on, it creates all instances at the same time.

I'd assume one might run into rate limits with huge setups, but for us that hasn't happened yet.

4censord committed:
feat: run instance creation in parallel

href commented:
Thank you for your PR. I took a look at it, and I think the direction is right, but there are some issues with the approach, due to backend constraints you could not be aware of:

The server API `DELETE` and `POST` calls are largely serial in the backend. If they arrive at the endpoint at the same time, they will mostly be worked on in a serialized fashion. This is partly an implementation detail, but also has to do with scheduling: The algorithm that places a VM in OpenStack considers one VM at a time, as scheduling multiple VMs in one call would be orders of magnituted harder from algorithm standpoint.

So if you were to time it, you would find that you do not gain much with concurrency.

What does matter is waiting for servers to be actually started (via the `WaitFor` call). So I suggest to update this PR as follows:

- Create servers serially, then await them concurrently. Maybe the main thread can create the servers in a loop, send the resulting server instances to workers via a channel, then wait for all the workers to await the server.
- Limit how many servers are created concurrently. Maybe if a channel is used to await the servers, the channel could be size-constrainted, so that the main loop would block if too many workers are already awaiting servers. I mean this mainly as a some kind of upper bound for sanity that should never be breached (e.g., I don't expect there to ever be more than 10 servers launched in parallel).
- Do not use concurrency for delete requests. As noted, those are serial anyway, and in our experience deleting a number of servers serially, vs. concurrently does not make a difference in time. So the additional code complexity there is not worth it in my opinion.

Let me know what you think.

4censord committed:
revert: dont use concurency for instance deletion

4censord commented:
So I've implemented the “create in serial, await in parallel” approach, and reverted the deletion to serial.

But, instead of having a precreated/limited number of awaiters, it dynamically creates goroutines to wait for servers becoming ready.

Mostly because I dislike having an upper bound in the plugin, as these features are already present as part of the GitLab runner configuration (`max_instances` combined with `scale_throttle.limit` and `scale_throttle.burst` settings. [1] [2] )

Also, some pipelines we currently have more than 10 jobs that run in parallel, so that would slow these down.

[1]: https://docs.gitlab.com/runner/configuration/advanced-configuration/#the-runnersautoscaler-section

[2]: https://docs.gitlab.com/runner/configuration/advanced-configuration/#the-runnersautoscalerscale_throttle-section )

href commented:
Thanks, I think the approach looks fair now. Setting an upper bound borders a bit on the paranoid, so I think you make a good argument. Since server creation is serial now, I don't foresee the API being a bottleneck either: Checking the server state is fairly fast.

I didn't do a functional test yet, as I would like to do that in one go, together with the other PR, once that is read.

href commented on provider.go:
```diff
@@ -287,16 +288,42 @@ func (g *InstanceGroup) Increase(
+	servers := make(chan string)
+	// Because concurency, we use a channel for collecting errors, as
+	// slice.append is not thread safe
+	errors_c := make(chan error)
```
You might want to consider using `sync.Mutex` instead, as I think it might make the implementation a bit more straight-forward. That being said, if you keep the channel, you have to close it, or the `range errors_c` expression below will never finish.

4censord replied:
you're right, that feels easier

href commented on provider.go:
```diff
@@ -287,16 +288,42 @@ func (g *InstanceGroup) Increase(
+
+	// Await all servers being ready
+	var wg sync.WaitGroup
+	go func(servers <-chan string) {
```
This goroutine needs a `wg.Add(1)` / `wg.Done()` combo as well, as you otherwise can run into the race condition where the first server has not been popped out of the channel yet, when `wg.Wait()` is called below (I actually observed this on a test host).

4censord replied:
done

href commented on provider.go:
```diff
@@ -315,26 +342,27 @@ func (g *InstanceGroup) Increase(
+	wg.Wait()
+	// and we explicitly close the channel, so the waiting dispatcher
+	// (that for loop that recieves from the channel) can exit cleanly
+	close(servers)
```
I think this should be closed as soon as the loop above is finished (probably necessary when you add the additional `wg.Add` call, but I think it's also good to close as soon as we do not expect new items to be added to a channel.

4censord replied:
done

href commented on provider.go:
```diff
@@ -315,26 +342,27 @@ func (g *InstanceGroup) Increase(
+		errs = append(errs, err)
 	}
 
 	return len(servers), errors.Join(errs...)
```
The length here is zero if I'm not mistaken. I think you need to track the created servers in a separate variable and then return that. I'm not sure what GitLab Fleeting really expects when there are errors (does it just ignore the length of the servers created), but I think the best we can do is report accurately on the number of servers that we actually awaited (the await call can technically fail if a server cannot be scheduled, causing the server to first be launched, then discarded).

4censord replied:
Should be resolved by switching it back to a slice+mutex instead of the channel

4censord replied:
You are right, so i've added a counter instead

href reviewed (changes requested):
I had some trouble getting this to work. It looks like in its current form the `Increase` call never returns.

4censord committed:
fix: Use mutex instead of channel magic

4censord committed:
fix: create instances in serial, wait in parallel

href commented:
Thank you. I'll go ahead and merge this PR, and the other one once the tests pass. Afterward I want to create a beta release and integrate it into https://github.com/cloudscale-ch/gitlab-runner. This will give me some additional testing mileage and would also make it possible for you to run this already, pending a proper release.

However, quay.io currently throws HTTP 500 whenever I push a container, so this is currently slowing me down a bit. I'll wait and see first if this resolves itself, before maying publishing it elsewhere. Either way, I'm aiming to have at least the beta release this week, and a proper release by next week.

In the meantime, thank you for your contributions and your patience 😅
